### ISSUE TITLE
nvidia-container-toolkit-1.0.0-2.x86_64.rpm obsoletes breaks installation of previous nvidia-container-runtime packages

### ISSUE THREAD

**Issue by Perdjesk** [NVIDIA/nvidia-container-runtime#69]
# Description

The introduction of `nvidia-container-toolkit` with RPM spec which obsoletes `nvidia-container-runtime-hook` prevents the installation of packages that are depending on `nvidia-container-runtime-hook`.

Notes:
- Introduced by https://github.com/NVIDIA/nvidia-container-runtime/commit/664bc75b382b293c444ff6fc7dd347ad8673b5ab 

/cc @RenaudWasTaken

# Steps to reproduce
Centos 7.6

1. Setup RHEL repository according to https://nvidia.github.io/nvidia-container-runtime/
2. Install a specific version of `nvidia-container-runtime` that depends on `nvidia-container-runtime-hook` 
 Example: 
 ```
 yum install nvidia-container-runtime-2.0.0-3.docker18.09.7.x86_64
 ```

# Expected behavior
`nvidia-container-runtime` is installed 

# Actual result
```
Resolving Dependencies
--> Running transaction check
---> Package nvidia-container-runtime.x86_64 0:2.0.0-3.docker18.09.7 will be installed
--> Processing Dependency: nvidia-container-runtime-hook < 2.0.0 for package: nvidia-container-runtime-2.0.0-3.docker18.09.7.x86_64
Package nvidia-container-runtime-hook is obsoleted by nvidia-container-toolkit, but obsoleting package does not provide for requirements
--> Finished Dependency Resolution
Error: Package: nvidia-container-runtime-2.0.0-3.docker18.09.7.x86_64 (nvidia-container-runtime)
           Requires: nvidia-container-runtime-hook < 2.0.0
           Available: nvidia-container-runtime-hook-1.2.1-1.x86_64 (nvidia-container-runtime)
               nvidia-container-runtime-hook = 1.2.1-1
           Available: nvidia-container-runtime-hook-1.3.0-1.x86_64 (nvidia-container-runtime)
               nvidia-container-runtime-hook = 1.3.0-1
           Available: nvidia-container-runtime-hook-1.4.0-1.x86_64 (nvidia-container-runtime)
               nvidia-container-runtime-hook = 1.4.0-1
           Available: nvidia-container-runtime-hook-1.4.0-2.x86_64 (nvidia-container-runtime)
               nvidia-container-runtime-hook = 1.4.0-2
 You could try using --skip-broken to work around the problem
 You could try running: rpm -Va --nofiles --nodigest
```

# Known workaround
Disable obsoletes dependency resolution by `yum`
```
yum --setopt=obsoletes=0 install nvidia-container-runtime-2.0.0-3.docker18.09.7.x86_64
```

# Discussions
If the intended behavior is to replace "in-place"  `nvidia-container-runtime-hook` for <3.0.1 releases then `nvidia-container-toolkit` must be marked has providing (`Provides: nvidia-container-runtime-hook`) the hook and with the version range for which the in-place replacement is wished for.

If the intended behavior is to keep <3.0.1 release to use `nvidia-container-runtime-hook` and only use `nvidia-container-toolkit` for >=3.0.1 releases then `Obsoletes: nvidia-container-runtime-hook` must not be included in the RPM spec of `nvidia-container-toolkit`

**Comment by RenaudWasTaken** [comment]
Ah, sorry about that!

Thanks for pointing to the `provides` directive, we'll release a new one very shortly!

**Comment by Perdjesk** [comment]
This is fixed with https://github.com/NVIDIA/nvidia-container-runtime/commit/fe70aec197bbf5202d806ed702a51974752e0510 
with the introduction of `Provides`:
```
# repoquery --provides nvidia-container-toolkit-1.0.1
nvidia-container-runtime-hook
nvidia-container-toolkit = 1.0.1-2
nvidia-container-toolkit(x86-64) = 1.0.1-2
```

Could please confirm the intended behavior is to replace all others package dependencies to `nvidia-container-runtime-hook` with `nvidia-container-toolkit`?

For example the installation of `nvidia-container-runtime-2.0.0-1.docker1.13.1.x86_64` is now resolving as follow:
```
Dependencies Resolved

==============================================================================================================================================================================================================================================================================
 Package                                                                 Arch                                                 Version                                                            Repository                                                              Size
==============================================================================================================================================================================================================================================================================
Installing:
 nvidia-container-runtime                                                x86_64                                               2.0.0-1.docker1.13.1                                               nvidia-container-runtime                                               1.6 M
Installing for dependencies:
 libnvidia-container-tools                                               x86_64                                               1.0.2-1                                                            libnvidia-container                                                     33 k
 libnvidia-container1                                                    x86_64                                               1.0.2-1                                                            libnvidia-container                                                     74 k
 nvidia-container-toolkit                                                x86_64                                               1.0.1-2                                                            nvidia-container-runtime                                               617 k

Transaction Summary
==============================================================================================================================================================================================================================================================================
Install  1 Package (+3 Dependent packages)
```

When is it expected to release the 3.X.X version and update `master` branch?

**Comment by RenaudWasTaken** [comment]
I've pushed the new release yesterday and will create the new release in a few minutes.

**Comment by RenaudWasTaken** [comment]
I've created the release!

Some details as to the issues we faced:
1. As you mentioned we didn't include the `Provide` directive which made the hook package obsolete but prevented users from upgrading to the toolkit
2. As part of the toolkit package, we renamed the binary `nvidia-container-runtime-hook` to `nvidia-container-toolkit`.
For backwards compatibility reasons, the toolkit was creating a symlink from the hook to the toolkit binary.
We however noticed that the way the upgrade path was handled by yum is that it would install the toolkit first and then remove the hook. This basically meant that we never had the opportunity to create the symlink and if we ever did, removing the hook package also removed the symlink.
We fixed this by creating the symlink at the end of the transaction (the `%posttrans` directive).
